File: src/functions/project.ts
import { invoke } from "@tauri-apps/api/tauri"
import { Project } from "../store/project/types"

export async function createProject(filepath: string): Promise<Project> {
    const result: any = await invoke("create_project", {
        audioFilepath: filepath,
    })
    if (result.status === "Success") {
        return result.project as Project
    } else {
        console.log("Unable to create project:", result)
        throw new Error("Unable to create project")
    }
}

export async function getAllProjects(): Promise<Project[]> {
    const result: any = await invoke("get_all_projects", {})
    if (result.status === "Success") {
        return result.projects as Project[]
    } else {
        console.log("Unable to get projects", result)
        throw new Error("Unable to get projects.")
    }
}

File: src-tauri/src/routes/project.rs
use std::path::PathBuf;
use tokio::sync::Mutex;

use serde::{self, Deserialize, Serialize};
use tauri::State;

use crate::data::{AppDb, Project};

use super::{Error, Result};

#[derive(Serialize, Deserialize)]
#[serde(tag = "status")]
pub enum CreateProjectResponse {
    #[serde(alias = "success")]
    Success { project: Project },
}

#[derive(Serialize, Deserialize)]
#[serde(tag = "status")]
pub enum GetAllProjectsResponse {
    #[serde(alias = "success")]
    Success { projects: Vec<Project> },
}

#[tauri::command]
pub async fn create_project(
    audio_filepath: &str,
    app_db_mutex: State<'_, Mutex<AppDb>>,
) -> Result<CreateProjectResponse> {
    let app_db = app_db_mutex.lock().await;

    app_db
        .create_project(PathBuf::from(audio_filepath))
        .map_or(Err(Error::ProjectCreationError), |project| {
            Ok(CreateProjectResponse::Success { project })
        })
}

#[tauri::command]
pub async fn get_all_projects(
    app_db_mutex: State<'_, Mutex<AppDb>>,
) -> Result<GetAllProjectsResponse> {
    let app_db = app_db_mutex.lock().await;
    app_db
        .get_projects()
        .map_or(Err(Error::ProjectCreationError), |projects| {
            Ok(GetAllProjectsResponse::Success { projects })
        })
}
